perceval/components/source.py

```diff
@@ -201,16 +181,48 @@ def generate_distribution(self, expected_input: BasicState, prob_threshold: floa
         :param prob_threshold: Probability threshold under which the resulting state is filtered out. By default,
             `global_params['min_p']` value is used.
         """
-        dist = SVDistribution()
         prob_threshold = max(prob_threshold, global_params['min_p'])
         get_logger().debug(f"Apply 'Source' noise model to {expected_input}", channel.general)
-        for photon_count in expected_input:
-            dist = SVDistribution.tensor_product(dist, self.probability_distribution(photon_count), prob_threshold)
+
+        distributions = [self.probability_distribution(photon_count, prob_threshold) for photon_count in expected_input]
+        dist = SVDistribution.list_tensor_product(distributions, prob_threshold=prob_threshold)
+
         dist.normalize()
         if self.simplify_distribution and self.partially_distinguishable:
             dist = anonymize_annotations(dist, annot_tag='_')
         return dist
 
+    def generate_samples(self, max_samples, expected_input: BasicState):
+        """
```

Comment:
It seems the doctring was copied from another function and does not correspond to this function